### PR TITLE
feat(operator): watch manipulated resources for reconcile on change

### LINKED ISSUE #618: Should react to changes in dependent Resources
##### What can be improved

The session should reconcile on a change to any resource we change.

##### Why

If some 3. party update e.g. a VirtualService outside of our control we can reapply what we need.

### PR DESCRIPTION
**Pull request by aslakknutsen**
#### Short description of what this resolves:

Any of the manipulated Resource can be changed externally at any time and break the Session, either via manual or redeploys scripts etc.

#### Changes proposed in this pull request:

Watch all manipulated Resources for possible change and reconcile if needed.

Fixes #618

### CODE REVIEW

File: pkg/reference/enqueue_annotations.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
+package reference
```
Comment (bartoszmajsak) [inline pkg/reference/enqueue_annotations.go:1]: I still wonder if we should leave the original source/license here

File: pkg/reference/enqueue_annotations.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+		logChange("Reference Object updated", evt.ObjectOld, reqs)
+		e.addToQueue(q, reqs)
+	}
+	if ok, reqs := e.getAnnotationRequests(evt.ObjectNew); ok {
+		logChange("Reference Object updated", evt.ObjectNew, reqs)
+		e.addToQueue(q, reqs)
+	}
+}
+
+// Delete implements EventHandler.
```
Comment (bartoszmajsak) [inline pkg/reference/enqueue_annotations.go:96]: Those are rather brief ;)

File: pkg/reference/enqueue_annotations_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			},
+		}
+	})
+
+	It("should fail to add when missing namespace", func() {
+		// when
+		err := reference.Add(types.NamespacedName{Name: "session1"}, deployment)
+		Expect(err).To((HaveOccurred()))
+	})
+	It("should fail to add when missing name", func() {
```
Comment (bartoszmajsak) [inline pkg/reference/enqueue_annotations_test.go:32]:
```suggestion

	It("should fail to add when missing name", func() {
```

**Review by bartoszmajsak — APPROVED** [review]
Few suggestions and comments :) Otherwise, let's make that k8s tests passing and it's ready to go. I will approve to not block from merging as I'm away today, but I trust you will at least consider what I wrote below ;)

**Comment by alien-ike** [comment]
@aslakknutsen: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-tests-openshift | 2a7faa6fb026ad53b661b36235ff4829bffc54d8 | [link](http://status-prow.apps.workspace.maistra.test/view/gs/ike-prow-logs/pr-logs/pull/maistra_istio-workspace/719/e2e-tests-openshift/1366344623743569920) | `/test openshift`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: controllers/session/session_controller.go [hunk 4]
```diff
@@ -269,8 +283,8 @@ func (r *ReconcileSession) sync(ctx model.SessionContext, session *istiov1alpha1
 		}
 	}
 	if located {
-		for _, mutator := range r.manipulators.Mutators {
-			err := mutator(ctx, ref)
+		for _, mutator := range r.manipulators.Handlers {
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply controllers/session/session_controller.go:294]: Resolved with fa716c2

File: e2e/smoke_test.go [hunk 5]
```diff
@@ -227,6 +227,60 @@ var _ = Describe("Smoke End To End Tests - against OpenShift Cluster with Istio
...
+			BeforeEach(func() {
+				scenario = "scenario-1"
+			})
+
+			It("should watch for changes in ratings service and serve it", func() {
+				EnsureAllDeploymentPodsAreReady(namespace)
+				EnsureProdRouteIsReachable(namespace, ContainSubstring("ratings-v1"), Not(ContainSubstring(PreparedImageV1)))
+
+				// when we start ike to create
+				ike1 := RunIke(tmpDir, "create",
```
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply e2e/smoke_test.go:241]: Resolved with e5c942e

**Comment by alien-ike** [comment]
@aslakknutsen: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-tests-openshift | e26ea55f45e6411b28ef12dec07535168933ea01 | [link](http://status-prow.apps.workspace.maistra.test/view/gs/ike-prow-logs/pr-logs/pull/maistra_istio-workspace/719/e2e-tests-openshift/1366369789638873088) | `/test openshift`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: pkg/reference/enqueue_annotations.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/reference/enqueue_annotations.go:96]: Resolved in 5cfeed0

File: pkg/reference/enqueue_annotations.go (hunk 1 above)
Comment (bartoszmajsak) [reply pkg/reference/enqueue_annotations.go:1]: Resolved in d98315b

**Pull request merged**